### PR TITLE
Fixed issue where IsNullable causing JSON issues

### PR DESCRIPTION
**Pull request by TwistedTwigleg**
*Description of changes:*

Changes how the ShadowState is converted to and from JSON to fix the issue where it gives incorrect (and non-backwards compatible) results when converting. For the overwhelming majority of cases, the ShadowState you convert to and from JSON will be exactly the same.

However, due to a limitation in the GSON library, there are cases where it is not 100% the same when converted to and from. With this change though, it is 100% backwards compatible with the functionality prior to the shadow nullable changes.

_________

The issue stems from how GSON converts to and from JSON. In Java, when a value is not defined it is set to `null`, so if the code converts the JSON to a ShadowState through GSON serialization, if we check to see if `reported` or `desired` is `null`, it will return true when it is set to `null` in the JSON and when it is not present in the JSON.

So I looked at crawling through the JSON file using the JSON reader provided by JSON, as then I could see if the `desired` or `null` fields are present and then set `IsNullable` accordingly. However, crawling the JSON didn't work as I would immediately get a `DOCUMENT_END` token because it was already crawled by `delegate.read`. Trying to copy the JSON Reader prior to crawling, trying to get the raw JSON string from JSON Reader, and more didn't work.

I also tried converting it to a JSON tree using `delegate.toJsonTree`, but unfortunately this did not work either, as checking to see if a value was `null` returned true if it was null in the JSON or if it was not present in the tree at all.

So, I couldn't figure it out. The only code path this affects is if you pass a JSON document with either `reported` or `desired` to `null`, as it will not set `IsNullable` to true for that variable. You can manually set this though in the ShadowState though if you need it to be null.

So now the code just defaults to not setting the `IsNullable` variables when reading from JSON. This gives 100% backwards compatibility and shouldn't cause unwanted surprises, it just unfortunately isn't 100% the same in all situations if you convert to and from JSON on the exact same ShadowState if the ShadowState has either `desired` or `reported` as `null`.

________

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.*

### CODE REVIEW

**Comment by TwistedTwigleg** [comment]
**Testing**

I added the following code to the Shadow Sample for testing after `shadow = new IotShadowClient(connection);` around line 240 below:

<details>
~~~
// =======================
// TMP - shadow tests
Gson shadowGson = shadow.getShadowGson();

// -------------------------------------------------------------------------------------
// FIRST TEST: Compare against hard-coded JSON
ShadowState test_state_one = new ShadowState();
test_state_one.desired = new HashMap<String, Object>() {{
    put("ThingOne", 10.0);
    put("ThingTwo", "Bob");
}};
String test_state_one_json = shadowGson.toJson(test_state_one);
String test_state_two_json = "{\"desired\":{\"ThingTwo\":\"Bob\",\"ThingOne\":10.0}}";

System.out.println("====== TEST 1: Hard coded JSON compare ======");
System.out.println("Test_state_one_json: " + test_state_one_json);
System.out.println("Test_state_two_json: " + test_state_two_json);
boolean isFirstTestEqual = test_state_one_json.equals(test_state_two_json);
System.out.println("Is equal result: " + isFirstTestEqual);
// -------------------------------------------------------------------------------------

// -------------------------------------------------------------------------------------
// SECOND TEST - compare JSON outputs
System.out.println("====== TEST 2: JSON from payload compared to payload (as close to) ======");
ShadowState test_state_two = shadowGson.fromJson(test_state_two_json, ShadowState.class);
String test_state_two_from_payload = shadowGson.toJson(test_state_two);

System.out.println("Test_state_one_json: " + test_state_one_json);
System.out.println("Test_state_two_from_payload_json: " + test_state_two_from_payload);
boolean isSecondTestEqual = test_state_one_json.equals(test_state_two_from_payload);
System.out.println("Is equal result: " + isSecondTestEqual);

// -------------------------------------------------------------------------------------
// THIRD TEST - send null through JSON
System.out.println("====== TEST 3: JSON from payload compared to payload (as close to) ======")
ShadowState test_state_three = new ShadowState();
test_state_three.desired = new HashMap<String, Object>() {{
    put("ThingOne", 10.0);
    put("ThingTwo", "Bob");
}};
test_state_three.reported = null;
test_state_three.reportedIsNullable = true;
String test_state_three_json = shadowGson.toJson(test_state_three);
String test_state_three_json_expected = "{\"desired\":{\"ThingTwo\":\"Bob\",\"ThingOne\":10.0},\"reported\":null}";

System.out.println("Test_state_three_json: " + test_state_three_json);
System.out.println("Test_state_three_expected_json: " + test_state_three_json_expected);
boolean isThirdTestEqual = test_state_three_json.equals(test_state_three_json_expected);
System.out.println("Is equal result: " + isThirdTestEqual);

// -------------------------------------------------------------------------------------
// FOURTH TEST - Parse with null
System.out.println("====== TEST 4: Parse JSON from payload with null ======");
ShadowState test_state_four = shadowGson.fromJson("{\"desired\":null, \"reported\":{\"hat\":\"fancy\"}}", ShadowState.class);
String test_state_four_json = shadowGson.toJson(test_state_four);
String test_state_four_json_expected = "{\"desired\":null,\"reported\":{\"hat\":\"fancy\"}}";

System.out.println("Test_state_four_json: " + test_state_four_json);
System.out.println("Test_state_four_expected_json: " + test_state_four_json_expected);
boolean isFourthTestEqual = test_state_four_json.equals(test_state_four_json_expected);
System.out.println("Is equal result: " + isFourthTestEqual);

System.exit(1);
// =======================
~~~
</details>

Note that in the tests above, the fourth test fails. This is due to the limitation mentioned above.

**Comment by TwistedTwigleg** [comment]
You also need to add the following to `IotShadowClient.java`:

~~~
public Gson getShadowGson() {
    return this.gson;
}
~~~

As otherwise you cannot convert the ShadowState objects to and from JSON directly for testing.

I didn't find any testing files in the repository, but I will look again and if I find the testing files, I'll update this PR accordingly.

**Comment by TwistedTwigleg** [comment]
Added tests that run automatically to make sure the ShadowState works as expected 👍

File: sdk/tests/iot/ShadowStateTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        }};
+        state.reported = null;
+        state.reportedIsNullable = true;
+        String state_json = shadowGson.toJson(state);
+        String expected_json = "{\"desired\":{\"ThingTwo\":\"Bob\",\"ThingOne\":10.0},\"reported\":null}";
+        assertEquals(state_json, expected_json);
+    }
+
+    /*
+    @Test
```
Comment (bretambrose) [inline sdk/tests/iot/ShadowStateTest.java:84]: delete?

**Review by bretambrose — APPROVED** [review]

**Comment by TwistedTwigleg** [comment]
Thanks for the review! Merging into main...

**Pull request merged**